File: api/src/main/scala/gg/kenken/KenKenSolver.scala
package gg.kenken

import scala.io.Source
import scala.language.reflectiveCalls
import scala.collection.mutable.HashMap
import scala.util.matching.Regex

import com.typesafe.scalalogging.LazyLogging

import gg.kenken.solver._


object Control {
  def using[A <: { def close(): Unit }, B](resource: A)(f: A => B): B =
    try {
      f(resource)
    } finally {
      resource.close()
    }
}

object KenKenSolver extends LazyLogging {
  def parseConstraintLine(constraintLine: String): List[Constraint] = {
    val constraintBlobPattern: Regex = "([^=]*)=([0-9]+)([\\+\\-x/\\.]?)".r

    constraintLine.split(" ").map(token => token match {
      case constraintBlobPattern(name, value, operator) => new Constraint(name(0), value.toInt, Operator.withName(operator))
      case _ => throw new KenKenSolverException("Error parsing token " + token)
    }).toList
  }

  def parseBoard(boardLines: Iterator[String]): Array[Array[Char]] =
    boardLines.map(line =>
      line.split(" ").map(charStr =>
        charStr(0)
      )
    ).toArray

  def parseFile(fileName: String): GameState = {
    Control.using(Source.fromFile(fileName)) { source =>
      var lines = source.getLines
      return GameState(parseConstraintLine(lines.next), parseBoard(lines))
    }
  }

  def solveFromAPI(boardStrings: Iterator[String], constraintString: String): String =
    GameState(parseConstraintLine(constraintString), parseBoard(boardStrings))
      .solve
      .toString

  def main(args: Array[String]): Unit = {
    var gameState = this.parseFile("input/kenken3.in")
    logger.info("{}", gameState.constraintsToString)
    logger.info("{}", gameState.boardToString)
    val theSolvedBoard = gameState.solve
    logger.info("SOLVED ಡ_ಡ")
    logger.info("{}", theSolvedBoard)
  }
}

File: web/src/solver/KenKenSolver.js
import { Constraint } from './Constraint';
import { operatorFromString } from './Operator';
import { KenKenSolverException } from './KenKenSolverException';
import { createGameState } from './GameState';

export function parseConstraintLine(constraintLine) {
  const constraintBlobPattern = /([^=]*)=([0-9]+)([\+\-x/\.]?)/;

  return constraintLine.split(' ').map(token => {
    const match = token.match(constraintBlobPattern);
    if (match) {
      const name = match[1].charAt(0);
      const value = parseInt(match[2], 10);
      const operator = operatorFromString(match[3]);
      return new Constraint(name, value, operator);
    } else {
      throw new KenKenSolverException('Error parsing token ' + token);
    }
  });
}

export function parseBoard(boardLines) {
  return boardLines.map(line =>
    line.split(' ').map(charStr => charStr.charAt(0))
  );
}

export function solveFromAPI(boardStrings, constraintString) {
  const constraints = parseConstraintLine(constraintString);
  const board = parseBoard(boardStrings);
  const gameState = createGameState(constraints, board);
  const solved = gameState.solve();
  if (solved) {
    return solved.toString();
  }
  throw new KenKenSolverException('No solution found');
}
